File: mpi_example6.py
from mpi4py import MPI
import numpy as np

def main():
    
    comm = MPI.COMM_WORLD
    my_rank = comm.Get_rank()
    nprocs = comm.Get_size()

    prev = my_rank - 1
    next = my_rank + 1

    if my_rank == 0:
        prev = nprocs - 1
    if my_rank == nprocs - 1:
        next = 0

    rbuf1 = np.empty(1, dtype=int)
    rbuf2 = np.empty(1, dtype=int)
    sbuf = np.array([my_rank], dtype=int)
    
    tag1 = 1
    tag2 = 2
    reqs = [None] * 4

    # non-blocking receives
    reqs[0] = comm.Irecv([rbuf1, MPI.INT], source=prev, tag=tag1)
    reqs[1] = comm.Irecv([rbuf2, MPI.INT], source=next, tag=tag2)
    # non-blocking sends
    reqs[2] = comm.Isend([sbuf, MPI.INT], dest=prev, tag=tag2)
    reqs[3] = comm.Isend([sbuf, MPI.INT], dest=next, tag=tag1)

    # can do some work here if you want, e.g. print current
    # state of the receive buffers
    print(rbuf1[0], rbuf2[0])
    
    MPI.Request.Waitall(reqs)
    print("My rank: {:d} and my neighbors: {:d} {:d}".format(my_rank, rbuf1[0], rbuf2[0]))
    
if __name__ == "__main__":
    main()

File: mpi_example6.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

int main(int argc, char *argv[]) {
  
  int numtasks, my_rank, prev, next;
  
  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &numtasks);
  MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
  
  prev = my_rank - 1;
  next = my_rank + 1;
  
  if (my_rank == 0) {
    prev = numtasks - 1;
  }
  if (my_rank == numtasks - 1) {
    next = 0;
  }

  int sbuf = my_rank;
  int rbuf1, rbuf2;
  
  MPI_Request reqs[4];

  // Non-blocking receives
  MPI_Irecv(&rbuf1, 1, MPI_INT, prev, 1, MPI_COMM_WORLD, &reqs[0]);
  MPI_Irecv(&rbuf2, 1, MPI_INT, next, 2, MPI_COMM_WORLD, &reqs[1]);
  // Non-blocking sends
  MPI_Isend(&sbuf, 1, MPI_INT, prev, 2, MPI_COMM_WORLD, &reqs[2]);
  MPI_Isend(&sbuf, 1, MPI_INT, next, 1, MPI_COMM_WORLD, &reqs[3]);

  // Can do some work here if needed
  // For example, print current state of the receive buffers
  printf("%d %d\n", rbuf1, rbuf2);
  
  // Wait for all non-blocking operations to complete
  MPI_Waitall(4, reqs, MPI_STATUSES_IGNORE);
  
  printf("My rank: %d and my neighbors: %d %d\n", my_rank, rbuf1, rbuf2);
  
  MPI_Finalize();
  return 0;
}
